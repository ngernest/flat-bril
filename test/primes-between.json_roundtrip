{
  "functions": [
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "label": "for.outer.init"
        },
        {
          "dest": "t0",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "a",
            "t0"
          ],
          "dest": "t1",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "t1"
          ],
          "funcs": [],
          "labels": [
            "true",
            "false"
          ],
          "op": "br"
        },
        {
          "label": "true"
        },
        {
          "dest": "t2",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.outer.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "false"
        },
        {
          "args": [
            "a"
          ],
          "dest": "t2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.outer.cond"
        },
        {
          "args": [
            "t2",
            "b"
          ],
          "dest": "t3",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "t3"
          ],
          "funcs": [],
          "labels": [
            "for.outer.body",
            "for.outer.end"
          ],
          "op": "br"
        },
        {
          "label": "for.outer.body"
        },
        {
          "dest": "t4",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "t4",
            "t4"
          ],
          "dest": "t5",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "label": "for.inner.init"
        },
        {
          "dest": "t6",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "t7",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "label": "for.inner.cond"
        },
        {
          "args": [
            "t2",
            "t7"
          ],
          "dest": "t8",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "t6",
            "t8"
          ],
          "dest": "t9",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "t9"
          ],
          "funcs": [],
          "labels": [
            "for.inner.body",
            "for.inner.end"
          ],
          "op": "br"
        },
        {
          "label": "for.inner.body"
        },
        {
          "args": [
            "t2",
            "t6"
          ],
          "dest": "t10",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "dest": "t11",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "t10",
            "t11"
          ],
          "dest": "t12",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "t12"
          ],
          "funcs": [],
          "labels": [
            "if.inner.body",
            "if.inner.end"
          ],
          "op": "br"
        },
        {
          "label": "if.inner.body"
        },
        {
          "dest": "t13",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "t14",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "t13",
            "t14"
          ],
          "dest": "t5",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.inner.end"
          ],
          "op": "jmp"
        },
        {
          "label": "if.inner.end"
        },
        {
          "dest": "t15",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "t15",
            "t6"
          ],
          "dest": "t6",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.inner.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "for.inner.end"
        },
        {
          "dest": "t16",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "t6",
            "t16"
          ],
          "dest": "t6",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "t5"
          ],
          "funcs": [],
          "labels": [
            "if.outer.body",
            "if.outer.end"
          ],
          "op": "br"
        },
        {
          "label": "if.outer.body"
        },
        {
          "args": [
            "t2"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "label": "if.outer.end"
        },
        {
          "dest": "t17",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "t2",
            "t17"
          ],
          "dest": "t2",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.outer.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "for.outer.end"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "t1",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "b",
            "t1"
          ],
          "dest": "t2",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a",
            "t2"
          ],
          "dest": "t3",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "t3"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
