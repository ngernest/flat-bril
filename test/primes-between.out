@main
.for.outer.init
	op: const		dest: t0		type: int		value: 2	
	op: lt   		dest: t1		type: bool		args: ["a", "t0"]	
	op: br   		args: ["t1"]		labels: ["true", "false"]
.true
	op: const		dest: t2		type: int		value: 2	
	op: jmp  		labels: ["for.outer.cond"]
.false
	op: id   		dest: t2		type: int		args: ["a"]	
.for.outer.cond
	op: le   		dest: t3		type: bool		args: ["t2", "b"]	
	op: br   		args: ["t3"]		labels: ["for.outer.body", "for.outer.end"]
.for.outer.body
	op: const		dest: t4		type: int		value: 1	
	op: eq   		dest: t5		type: bool		args: ["t4", "t4"]	
.for.inner.init
	op: const		dest: t6		type: int		value: 2	
	op: const		dest: t7		type: int		value: 2	
.for.inner.cond
	op: div  		dest: t8		type: int		args: ["t2", "t7"]	
	op: le   		dest: t9		type: bool		args: ["t6", "t8"]	
	op: br   		args: ["t9"]		labels: ["for.inner.body", "for.inner.end"]
.for.inner.body
	op: call 		dest: t10		type: int		args: ["t2", "t6"]		funcs: ["mod"]
	op: const		dest: t11		type: int		value: 0	
	op: eq   		dest: t12		type: bool		args: ["t10", "t11"]	
	op: br   		args: ["t12"]		labels: ["if.inner.body", "if.inner.end"]
.if.inner.body
	op: const		dest: t13		type: int		value: 1	
	op: const		dest: t14		type: int		value: 2	
	op: eq   		dest: t5		type: bool		args: ["t13", "t14"]	
	op: jmp  		labels: ["for.inner.end"]
.if.inner.end
	op: const		dest: t15		type: int		value: 1	
	op: add  		dest: t6		type: int		args: ["t15", "t6"]	
	op: jmp  		labels: ["for.inner.cond"]
.for.inner.end
	op: const		dest: t16		type: int		value: 1	
	op: add  		dest: t6		type: int		args: ["t6", "t16"]	
	op: br   		args: ["t5"]		labels: ["if.outer.body", "if.outer.end"]
.if.outer.body
	op: print		args: ["t2"]	
.if.outer.end
	op: const		dest: t17		type: int		value: 1	
	op: add  		dest: t2		type: int		args: ["t2", "t17"]	
	op: jmp  		labels: ["for.outer.cond"]
.for.outer.end
@mod
	op: div  		dest: t1		type: int		args: ["a", "b"]	
	op: mul  		dest: t2		type: int		args: ["b", "t1"]	
	op: sub  		dest: t3		type: int		args: ["a", "t2"]	
	op: ret  		args: ["t3"]	
